### DISCUSSION TITLE
processor for array or json objects

### DISCUSSION THREAD

**Discussion by davidwormald** [redpanda-data/connect#2008] [General]
Hi, I need to have an output that is a json array of objects.  I tried using the awk processor and, of course, found that it created an array of strings :-)
My input does not have an array of objects so I cannot do a simple transform on the input.

How can I create an output that looks like this json:
[ 
   { "a": 1, "b": 2},
   { "a": 3, "b": 4}
]

instead of: 
[ 
   "{ \"a\": 1,\ "b\": 2}",
   "{ \"a\": 3, \"b\": 4}"
]

Thanks!

**Comment by mihaitodor** [comment]
@davidwormald I'm not sure I follow how you're getting that single `\` character in your output, since that looks like some invalid JSON. What I'd expect is to see `\` in front of every `"`. If that's the case, then you can use the bloblang [`parse_json()` method](https://www.benthos.dev/docs/guides/bloblang/methods#parse_json). However, since you mentioned the `awk` processor, have you considered implementing that transformation in pure bloblang? It might be easier to maintain it that way

**Reply by davidwormald** [reply]
Cheers. Yes,  I mistyped the JSON.  Let me copy/paste.
This is what I have - an array of strings:
{
	"fields": [
		"{\"attributes\":\"masked\",\"fieldType\":\"string\",\"label\":\"SSN\",\"name\":\"ssn\"}",
		"{\"attributes\":\"bold\",\"fieldType\":\"number\",\"label\":\"score\",\"name\":\"Score\"}"
	],
	"version": 1
}

This is what I need to create:
{
	"fields": [
		{"attributes": "masked","fieldType": "string","label": "SSN","name":"ssn"}",
		{"attributes":  "bold","fieldType": "number","label": "score","name":"score"}
	],
	"version": 1
}
My original idea was to use bloblang but I'm not seeing how to create an array in bloblang.  Any hints?

**Reply by davidwormald** [reply]
OK, a combination of for_each and parse_json worked.
Thanks!

**Reply by mihaitodor** [reply]
Note that what you shared as input is invalid JSON, so not sure if that's what you actually have to deal with. However, assuming your input is this:
```json
{
"fields": [
"{\"attributes\":\"masked\",\"fieldType\":\"string\",\"label\":\"SSN\",\"name\":\"ssn\"}",
"{\"attributes\":\"bold\",\"fieldType\":\"number\",\"label\":\"score\",\"name\":\"Score\"}"
],
"version": 1
}
```

Then you can use the following bloblang expression:
```coffee
root = this.assign({"fields": this.fields.map_each(f -> f.parse_json())})
```

**Reply by davidwormald** [reply]
I meant that my awk processor was creating the top json and what I wanted is in the second json.  The actual input has no array in it at all so I cannot map input to output; instead, I  need to create my own array.  I am using the mapping after I create an array in awk but I'm thinking there must be a better way.

**Reply by mihaitodor** [reply]
What does your initial input look like? Bloblang is perfectly capable of parsing anything really (see the `content()` function), but it's also fine to use the `awk` processor to parse the input if that's easier for you. However, having it emit non-valid JSON is a bit dodgy / hard to maintain.

**Reply by davidwormald** [reply]
Maybe I'm over-complicating my description.  Sorry. 
What I have to do is really simple. 

I'm not having a problem parsing input, but for a specific set input value I have to output valid JSON and somewhere in that JSON output is an array  of JSON objects and each object in the array has a couple of fields.    

I have it working as a two step thing:
1) use awk to build an array of string representations of objects 
2) then bloblang to to map each string to an actual object.

Wondering if there is an easier way.

**Reply by mihaitodor** [reply]
OK, yeah. So that's why I'm asking what is the input to your `awk` processor. I just realised that I should've answered your original question and I didn't:

> My original idea was to use bloblang but I'm not seeing how to create an array in bloblang. Any hints?

There are multiple ways, depending on what the input looks like. If it's a static array, you just do `root = ["foo", "bar", "baz"]`. If you need to do something like a for loop, the `range()` function might help: `root = range(0, 10, 2).map_each(v -> {"val": v})`. If you get some string as input that you need to split somehow, you can do this: `root = content().split(",")`. Sky's the limit :) Try using `benthos blobl server` to open a bloblang repl in your browser that you can use to quickly prototype the code.

**Comment by davidwormald** [comment]
That's perfect! Thanks!
